internal/validation/ingress/ingress.go

```diff
@@ -0,0 +1,81 @@
+package ingress
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/kong/go-kong/kong"
+	"github.com/sirupsen/logrus"
+	netv1 "k8s.io/api/networking/v1"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/failures"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/versions"
+	kongv1alpha1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1alpha1"
+)
+
+type routeValidator interface {
+	Validate(context.Context, *kong.Route) (bool, string, error)
+}
+
+func ValidateIngress(
+	ctx context.Context,
+	routesValidator routeValidator,
+	parserFeatures parser.FeatureFlags,
+	kongVersion semver.Version,
+	ingress *netv1.Ingress,
+) (bool, string, error) {
+	discardLogger := logrus.New()
+	discardLogger.Out = io.Discard
+	failuresCollector, err := failures.NewResourceFailuresCollector(discardLogger)
+	if err != nil {
+		return false, "", err
+	}
+	var icp kongv1alpha1.IngressClassParametersSpec
+	if kongVersion.LT(versions.ExplicitRegexPathVersionCutoff) {
+		icp.EnableLegacyRegexDetection = true
+	}
+	result := parser.IngressesV1ToKongServices(
+		parserFeatures,
+		[]*netv1.Ingress{ingress},
+		icp,
+		&parser.ObjectsCollector{}, // It's irrelevant for validation.
+		failuresCollector,
+	)
+
+	var errMsgs []string
+	for _, f := range failuresCollector.PopResourceFailures() {
+		errMsgs = append(errMsgs, f.Message())
+	}
+	if len(errMsgs) > 0 {
+		return false, validationMsg(errMsgs), nil
+	}
+	var kongRoutes []kong.Route
+	for _, r := range result {
+		for _, route := range r.Routes {
+			kongRoutes = append(kongRoutes, route.Route)
```

Comment:
Why not avoid allocating a separate slice for all routes and just call the `routesValidator.Validate` here?

Reply:
Yeah it could be done like that, but with a suggestion of providing a helper for parsing ingress, I lean toward returning routes, directly, I'll write it, you'll see 🙂

Reply:
This is my suggestion, I'm allocating, but on the other hand make it easier to read now imo